nxc/protocols/ssh.py

```diff
@@ -63,124 +79,220 @@ def create_conn_obj(self):
-            self.admin_privs = True
-            return True
-        stdin, stdout, stderr = self.conn.exec_command("sudo -ln | grep 'NOPASSWD: ALL'")
-        if stdout.read().decode("utf-8").find("NOPASSWD: ALL") != -1:
-            self.logger.info(f"Determined user is root via `sudo -ln` command")
-            self.admin_privs = True
-            return True
+        self.logger.info("Determined user is root via `id; sudo -ln` command")
+        _, stdout, _ = self.conn.exec_command("id; sudo -ln 2>&1")
+        stdout = stdout.read().decode(self.args.codec, errors="ignore")
+        admin_flag = {
+            "(root)": [True, None],
+            "NOPASSWD: ALL": [True, None],
+            "(ALL : ALL) ALL": [True, None],
+            "(sudo)": [False, f'Current user: "{self.username}" was in "sudo" group, please try "--sudo-check" to check if user can run sudo shell']
+        }
+        for keyword in admin_flag.keys():
+            match = re.findall(re.escape(keyword), stdout)
+            if match:
+                self.logger.info(f'User: "{self.username}" matched keyword: {match[0]}')
+                self.admin_privs = admin_flag[match[0]][0]
+                if not self.admin_privs:
+                    tips = admin_flag[match[0]][1]
+                else:
+                    break
+        if not self.admin_privs and "tips" in locals():
+            self.logger.display(tips)
+        return
+
+    def check_if_admin_sudo(self):
+        if not self.password:
+            self.logger.error("Check admin with sudo not support private key.")
+            return
+
+        if self.args.sudo_check_method:
+            method = self.args.sudo_check_method
+            self.logger.info(f"Doing sudo check with method: {method}")
+
+        if method == "sudo-stdin":
+            _, stdout, _ = self.conn.exec_command("sudo --help")
+            stdout = stdout.read().decode(self.args.codec, errors="ignore")
+            # Read sudo help docs and find "stdin"
+            if "stdin" in stdout:
+                shadow_Backup = f'/tmp/{uuid.uuid4()}'
+                # sudo support stdin password
+                self.conn.exec_command(f"echo {self.password} | sudo -S cp /etc/shadow {shadow_Backup} >/dev/null 2>&1 &")
+                self.conn.exec_command(f"echo {self.password} | sudo -S chmod 777 {shadow_Backup} >/dev/null 2>&1 &")
+                tries = 1
+                while True:
+                    self.logger.info(f"Checking {shadow_Backup} if it existed")
+                    _, _, stderr = self.conn.exec_command(f'ls {shadow_Backup}')
+                    if tries >= self.args.get_output_tries:
+                        self.logger.info(f'{shadow_Backup} not existed, maybe the pipe has been hanged over, please increase the number of tries with the option "--get-output-tries" or change other method with "--sudo-check-method". If it\'s still failing maybe sudo shell is not working with current user')
```

Comment:
Should be f-string and recommended English fix:
`self.logger.info(f"The file {shadow_backup} does not exist, the pipe may be hanging. Increase the number of tries with the option "--get-output-tries" or change other method with "--sudo-check-method". If it's still failing, maybe sudo shell does not work with the current user")`

Reply:
Fixed!